### DISCUSSION TITLE
eLockViolation exception when creating table from context command

### DISCUSSION THREAD

**Discussion by kwollaston** [CEXT-Dan/PyRx#97] [Q&A] [answered]
Hi, I am receiving an exception when trying to add agenerated table to the model, but only when executed from a context menu. The command works as expected when it is run in the command line and the table is added.

However, I am running the PyRxCmd_ function through a context menu command, and it does run the function and prints the expected table data in the console, but when it comes time to creating the table, it fails with the below exception

`Exception!(eLockViolation), function AcDbBlockTableRecord * openAcDbObject< AcDbBlockTableRecord>( PyDbObjectId ,enum AcDb::OpenMode,bool), Line 67, File PyDbObjectId.h:`

The offending code is:
```python
db = Db.curDb()
[...]
model = Db.BlockTableRecord(db.modelSpaceId(), Db.OpenMode.ForWrite)
model.appendAcDbEntity(table)
``` 

Why does the command line execution of the same function work but context menu execution does not?
I'm sure there is some nuance that I am missing, the context menu example doesn't do any writing, so I don't know if I need to do anything special. 

I have a feeling it has to do with the `db = Db.curDb()` line...

**Comment by gswifort** [comment]
Try this:
```py
lock = Ap.AutoDocLock()  # !!!
table = ...
db.addToModelspace(table)
del(table)
del(lock)
```

**Reply by kwollaston** [reply]
> Try this:
> 
> ```python
> lock = Ap.AutoDocLock()  # !!!
> table = ...
> db.addToModelspace(table)
> del(table)
> del(lock)
> ```

Thank you for the quick response! This did solve the issue. However, I would like to return the table from the function and I can't if it is deleted. If I comment out del(table), then it doesn't work properly and the table has no data.

Can you elaborate on how/why this works?

**Reply by gswifort** [reply, marked as answer]
```py
def func():
    lock = Ap.AutoDocLock() 
    table = ...
    id = db.addToModelspace(table)
    return id

id = func()
table = Db.Table(id)
```

@CEXT-Dan will probably explain better why it works this way.

**Comment by CEXT-Dan** [comment]
It’s application (session) vs command context 

If you’re not running the code from a command, i.e. from a modeless dialog or a menu, its more than likely that you’re in an application context. See[ AcApDocManager::lockDocument
](https://help.autodesk.com/view/OARX/2023/ENU/?guid=OARX-RefGuide-AcApDocManager__lockDocument_AcApDocument__AcAp__DocLockMode_ACHAR__ACHAR__bool)
In an application context, you may have database objects open for read without a lock, its’s the writing part that requires a lock.
You may use table.downgradeOpen() / table.upgradeOpen() to toggle the state, I.e. for returning your table from a function. 

As a side note, the default command (PyRxCmd_) is MODAL, however, you may create commands that work in a session context, see, testCmdFlags.py

also
```Python
lock = Ap.AutoDocLock() 
```

is short for

```Python
        doc = Ap.curDoc()
        docman = Ap.DocManager()
        docman.lockDocument(doc)
       
        #work on the database
       
        docman.unlockDocument(doc)
```
